### DISCUSSION TITLE
Closed enums

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#6319] [Language Ideas] [answered]
I decided to create a new discussion instead of having a few posts in other threads to centralize this topic to one place. Basically, my idea is to add a keyword, `closed` or `strict`, or something like that, which changes an `enum` to start prohibiting the conversion of an invalid backing value to said `enum`.
```cs
closed enum HorizontalAlignment
{
  Left = 1, Center = 2, Right = 3
}
var x = (HorizontalAlignment)0; // runtime error
// the above line is compiled to:
var x = Enum.ConvertToClosedEnum<HorizontalAlignment>(0); // this helper method would check and throw, if needed
x = (HorizontalAlignment)1; // OK
```
`[Flags]` can be supported by creating a pool of all possible values by bit-wise OR - it's just a set of numbers and not much different from the options' numbers themselves.
Implementing this can be done in a few way. One approach is to generate the set of possible values and the logic at runtime using IL emitting. Another would be generating a hidden static method at compile time and linking to said method with a hidden attribute on the enum type. There may be other hybrid approaches, as well.

**Comment by iam3yal** [comment, marked as answer]
Proposal #3179

**Reply by TahirAhmadov** [reply]
Oh my, I even commented in that issue and totally forgot about it :) Thank you.
